topics/integrate-in-existing-app.md

```diff
@@ -1,242 +1,415 @@
+
+   ![Warnings about platform-dependent code](warnings-android-specific-code.png){width=450}
+
+4. Remove Android-specific code by replacing it with cross-platform Kotlin code or connecting to Android-specific APIs using [`expect` and `actual` declarations](connect-to-platform-specific-apis.md). See the following sections for details.
+
+#### Replace Android-specific code with cross-platform code {initial-collapse-state="collapsed"}
+
+To make your code work well on both Android and iOS, replace all JVM dependencies with Kotlin dependencies wherever possible.
+
+1.  In the `login()` function of the `LoginDataSource` class, replace `IOException`, which is not available in Kotlin, with `RuntimeException`.
+
+    ```kotlin
+    // Before
+    return Result.Error(IOException("Error logging in", e))
+    ```
+
+    ```kotlin
+    //After
+    return Result.Error(RuntimeException("Error logging in", e))
+    ```
+
+2. For email validation, replace the `Patterns` class from the `android.utils` package with a Kotlin regular expression matching the pattern in the `LoginDataValidator` class:
+
+    ```kotlin
+    // Before
+    private fun isEmailValid(email: String) = Patterns.EMAIL_ADDRESS.matcher(email).matches()
+    ```
+
+    ```kotlin
+    // After
+    private fun isEmailValid(email: String) = emailRegex.matches(email)
+    
+    companion object {
+       private val emailRegex =
+               ("[a-zA-Z0-9\\+\\.\\_\\%\\-\\+]{1,256}" +
+                       "\\@" +
+                       "[a-zA-Z0-9][a-zA-Z0-9\\-]{0,64}" +
+                       "(" +
+                       "\\." +
+                       "[a-zA-Z0-9][a-zA-Z0-9\\-]{0,25}" +
+                       ")+").toRegex()
     }
-}
+    ```
+
+#### Connect to platform-specific APIs from the cross-platform code {initial-collapse-state="collapsed"}
+
+A universally unique identifier (UUID) for `fakeUser` in `LoginDataSource` is generated using the `java.util.UUID` class, which is not available for iOS.
+
+```kotlin
+val fakeUser = LoggedInUser(java.util.UUID.randomUUID().toString(), "Jane Doe")
 ```
 
-Note the `displayModel` and `lastModel` mechanism, which allows a view to detach and re-attach, for example for configuration changes on Android.
+Since the Kotlin standard library doesn't provide functionality for generating UUIDs, you still need to use platform-specific functionality for this case.
+
+Provide the `expect` declaration for the `randomUUID()` function in the shared code and its `actual` implementations for each platform – Android and iOS – in the corresponding source sets.
+You can learn more about [connecting to platform-specific APIs](connect-to-platform-specific-apis.md).
+
+1. Remove the `java.util.UUID` from the common code:
+
+    ```kotlin
+   val fakeUser = LoggedInUser(randomUUID(), "Jane Doe") 
+   ```
+
+1. Create a `Utils.kt` file in the `shared/src/commonMain` and provide the `expect` declaration:
```

Comment:
minor: in the `share/src/commonMain` _what?_ (folder/directory)

Reply:
Updated in all three places.